### PR TITLE
New SliceFlipNormalizeGPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**New feature** / **Refactoring**

## Description:

- It adds a new SliceFlipNormalize that doesn't employ flat indices and division by strides
- Limited to 2D for now
- Kernel will be used in CropMirrorNormalize
- Test data (input and reference) are generated in Python in in https://github.com/NVIDIA/DALI_extra/pull/113

### Affected modules and functionalities:

- New kernel

### Key points relevant for the review:

Kernel implementation

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3069

### CODE REVIEW

File: dali/kernels/slice/slice_flip_normalize_gpu.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * @tparam spatial_ndim Number of spatial dimensions: 2, 3, or 4 (max is seq of 3D)
+ * @tparam channel_dim: points to the channel dimension. -1 if no channel dimension
+ *
+ * @details see SliceFlipNormalizeGPU::Args
+ */
+template <typename Out, typename In, int spatial_ndim, int channel_dim>
+class DLL_PUBLIC SliceFlipNormalizeGPU {
+ public:
+  static_assert(channel_dim == 0 || channel_dim == spatial_ndim,
+                "Only channel first or channel last are supported");
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu.h:46]: True, channel_dim==-1 is not supported anymore. I'll update the doc.

File: dali/kernels/slice/slice_flip_normalize_gpu.cu [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    const Args &args) {
+  using Tile = kernels::BlockDesc<spatial_ndim>;
+  using Sample = SampleDesc<Out, In, spatial_ndim>;
+  int nsamples = in.num_samples();
+
+  Sample *samples_cpu = ctx.scratchpad->AllocatePinned<Sample>(nsamples);
+
+  float *norm_add_gpu = nullptr, *norm_mul_gpu = nullptr;
+  Out *fill_values_gpu = nullptr;
+  std::tie(norm_add_gpu, norm_mul_gpu, fill_values_gpu) = SetupParams(ctx, args);
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu.cu:148]: can't compile that

File: dali/kernels/slice/slice_flip_normalize_gpu.cu [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    c = 0;
+    for (; c < fill_values_sz; c++)
+      fill_values_data[c] = ConvertSat<Out>(a.fill_values[c]);
+    for (; c < out_nchannels_; c++)
+      fill_values_data[c] = ConvertSat<Out>(0.0f);
+  }
+
+  float *norm_add_gpu = nullptr, *norm_mul_gpu = nullptr;
+  Out *fill_values_gpu;
+  std::tie(norm_add_gpu, norm_mul_gpu, fill_values_gpu) = ctx.scratchpad->ToContiguousGPU(
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu.cu]: doesn't seem to work.

[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu.cu]: it does, I just had to rebase. Thanks

File: dali/kernels/slice/slice_flip_normalize_gpu.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu.cu]: Done now

**Review by klecki — APPROVED** [review]

File: dali/kernels/slice/slice_flip_normalize_gpu_test.cu [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    for (int i = 0; i < iters; i++)
+      kernel.Run(ctx, out_view, in_view, args);
+    CUDA_CALL(cudaEventRecord(finished, ctx.gpu.stream));
+    CUDA_CALL(cudaDeviceSynchronize());
+    float time_ms = 0;
+    CUDA_CALL(cudaEventElapsedTime(&time_ms, started, finished));
+    time_ms /= iters;
+    int64_t data_size =  // note: using out_view.num_elements() twice in purpose (we are not reading
+                        // all input)
+        out_view.num_elements() * sizeof(InType) + out_view.num_elements() * sizeof(OutType);
```
Comment (mzient) [inline dali/kernels/slice/slice_flip_normalize_gpu_test.cu:367]: Nitpick - weird formatting.
```suggestion
    // note: using out_view.num_elements() twice in purpose (we are not reading all input)
    int64_t data_size =  out_view.num_elements() * (sizeof(InType) + sizeof(OutType));
```

Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_gpu_test.cu]: Fixed

**Review by mzient — APPROVED** [review]

**Pull request merged**